DESIGN.md

```diff
@@ -0,0 +1,37 @@
+# Apollo Server design goals
+
+The goal of this project is to build a GraphQL server for Node.js that is simple, flexible, and performant. It is a Node.js GraphQL server built for production use.
+
+Apollo Server consists of three parts:
+
+1. Core
+2. Integrations
+3. Extensions
+
+At the core of Apollo Server is a function called `runQuery`, which handles parsing, validating and executing queries. Its interface is generic in order to allow for integrations with different Node.js server frameworks. Extensions provide useful functionality that can be shared between different integrations.
+
+
+### Core
+
+The main goals of Apollo Server are (in order of priority):
+
+1. Simplicity: Apollo Server’s core API is very straight forward. It’s one function that does one thing really well (parsing, validating and executing GraphQL queries), and doesn’t do anything else.
+2. Flexibility: The core of Apollo Server should be transport-agnostic (e.g. it doesn’t deal with HTTP or Websockets directly. This is will be handled in the wrappers for Express, HAPI, etc.)
+3. Performance: Apollo server should be be tunable to make it fast in production. One example of this is that it should be able to take pre-stored queries to skip parsing and validation. It should also allow easy integration of profiling tools like Apollo Tracer that help with debugging and optimizing server performance.
+
+### Integrations
+
+Apollo Server should come with a set of integrations for different Node.js server frameworks:
+
+- Express
+- HAPI
+- Connect
+- Koa
+- ...
+
+Framework integrations take care of parsing requests, submitting them to Apollo Server’s core runQuery  function, and sending the response back to the client. As a minimum requirement, wrappers have to accept requests over HTTP and pass the tests written for express-graphql. They should however be written in such a way that makes it easy to add features, such as websocket support, batched queries, subscriptions etc.
+
+Integrations should be limited to handling transport-specific (eg. setting headers) and framework-specific things (eg. registering a route).
+
+### Extensions
+Things that are not part of runQuery’s tasks, but are GraphQL specific (such as serving the GraphiQL UI, generating a schema, storing prepared queries, etc.)should be implemented in another core module of Apollo Server that lives alongside runQuery, or be imported from graphql-tools or other related packages.
```

Comment:
This doesn't seem fully fleshed out - I think putting in a bit of thought about what extensions might exist would pay off down the road. Btw serving GraphiQL over HTTP doesn't seem like a graphql specific thing, IMO that would go in the integration since you can't serve graphiql over a web socket for example.